### PR TITLE
Heap-allocate two locks, to avoid spurious errors when app is terminated

### PR DESCRIPTION
**Pull request by austinzheng**

### CODE REVIEW

File: Realm/RLMRealm.mm [hunk 1]
```diff
@@ -287,8 +287,12 @@ + (instancetype)realmWithConfiguration:(RLMRealmConfiguration *)configuration er
     realm->_dynamic = dynamic;
 
     // protects the realm cache and accessors cache
-    static std::mutex initLock;
-    std::lock_guard<std::mutex> lock(initLock);
+    static std::mutex *initLock;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        initLock = new std::mutex();
+    });
```
Comment (bdash) [inline Realm/RLMRealm.mm]: This can all be simplified to `static std::mutex& initLock = *new std::mutex;`

File: Realm/RLMRealmUtil.mm [hunk 2]
```diff
@@ -36,11 +36,15 @@
 #import <unistd.h>
 
 // Global realm state
-static std::mutex s_realmCacheMutex;
+static dispatch_once_t s_realmCacheMutexDispatchToken;
+static std::mutex *s_realmCacheMutex;
```
Comment (bdash) [inline Realm/RLMRealmUtil.mm]: Ditto about this being `static std::mutex& s_realmCacheMutex = *new std::mutex;`.

Note that heap-allocating this mutex without doing similarly for the data it protects (`s_realmsPerPath`) won't be super helpful.

File: Realm/RLMRealm.mm (hunk 1 above)
Comment (austinzheng) [inline Realm/RLMRealm.mm]: Oh, this is good to know. Thanks!

**Comment by umarniz** [comment]
Hi, I am coming from this issue:

https://github.com/realm/realm-cocoa/issues/4702

That references to this patch being a possible solution so firstly, thank you for the patch :).

Just wanted to point out that in the patch you are allocating the pointer on the heap to a static object but not freeing it anywhere and this would cause a memory leak, albeit only a tiny one (single mutex object = 64 bytes), as the heap is never de-allocated.

I could live with that as long as it resolves the realm crash I mentioned above as it is affecting ~50% of our users :(

**Comment by bdash** [comment]
There is no leak here. The heap-allocated objects are always reachable and in use until the point the application exits.

**Comment by umarniz** [comment]
@bdash sorry you are correct I just re-evaluated the condition and the heap allocated for the static object would be freed up with the application heap on exit.

I was using [leaker](http://left404.com/leaker/) to do leak tests but it would report this condition as a memory leak as we don't free the memory directly but it does free itself correctly for a static object when the application exits.

**Comment by bdash** [comment]
Correct, all memory within the application's address space is discarded when the process exits. Tools that attempt to detect leaks by matching `free` with `malloc` will report false-positive in cases like this.

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,7 +11,8 @@ x.x.x Release notes (yyyy-MM-dd)
 
 ### Bugfixes
 
-* None.
+* Fix an issue where spurious crashes might occur when the host application is
```
Comment (bdash) [inline CHANGELOG.md]: It's not accurate to say the crashes are spurious. They're real crashes.

I'd phase this more like:
> Fix a crash that could occur if new Realm instances were created while the application was exiting.

**Pull request merged**